source/simulator/assemblers/newton_stokes.cc

```diff
@@ -438,6 +455,43 @@ namespace aspect
 #endif
     }
 
+    template <int dim>
+    void
+    NewtonStokesIncompressibleTerms<dim>::
+    create_additional_material_model_outputs(MaterialModel::MaterialModelOutputs<dim> &outputs) const
+    {
+      const unsigned int n_points = outputs.viscosities.size();
+
+      if (this->get_parameters().enable_additional_stokes_rhs
+          && outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >() == nullptr)
+        {
+          outputs.additional_outputs.push_back(
+            std_cxx14::make_unique<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim>> (n_points));
+        }
+
+      Assert(!this->get_parameters().enable_additional_stokes_rhs
+             ||
+             outputs.template get_additional_output<MaterialModel::AdditionalMaterialOutputsStokesRHS<dim> >()->rhs_u.size()
+             == n_points, ExcInternalError());
+
+      if ((this->get_parameters().enable_elasticity) &&
+          outputs.template get_additional_output<MaterialModel::ElasticOutputs<dim> >() == nullptr)
+        {
+          outputs.additional_outputs.push_back(
+            std_cxx14::make_unique<MaterialModel::ElasticOutputs<dim>> (n_points));
+        }
+
+      Assert(!this->get_parameters().enable_elasticity
+             ||
+             outputs.template get_additional_output<MaterialModel::ElasticOutputs<dim> >()->elastic_force.size()
+             == n_points, ExcInternalError());
+
+      if (this->get_newton_handler().parameters.newton_derivative_scaling_factor == 0)
+        return;
+
+      NewtonHandler<dim>::create_material_model_outputs(outputs);
```

Comment:
Looking again at this code, I think it can be simplified into the following:

```suggestion
      if (this->get_newton_handler().parameters.newton_derivative_scaling_factor != 0)
        NewtonHandler<dim>::create_material_model_outputs(outputs);
```

Reply:
👍Aside from this change, are you satisfied with the current structure? There may be a way to get around some of the code duplication in the last commit, but with the change above it is just two lines being duplicated.

Reply:
The code in the default implementation in `NewtonInterface` can be deleted now.

The structure look correct to me. Although I think it is alright, I am just doubting whether this is the correct place to add the newton additional material model outputs. Any opinions? @gassmoeller @tjhei @jdannberg

Reply:
@MFraters - are operations other than `create_additional_material_model_outputs` performed in the `NewtonInterface` class? If not, can that class be removed? I agree there might be a better place for all of this code, although this is now consistent with the structure in the `stokes.cc` assembly.

Reply:
I guess you are right that it can be completely removed. Just replace the `:public NewtonInterface<dim>` with `: public aspect::Assemblers::Interface<dim>, public SimulatorAccess<dim>` where needed. 

Yes, you are right that in this way it conforms to `stokes.cc`, so I think the new structure is good.

Reply:
I was wrong, sorry. The `NewtonStokesPreconditioner` still needs to access the derivatives through the `NewtonInterface` (local tests failed). I vote to just keep the structure in the most recent commit unless someone has strong feelings about reorganization. I'll revisit this when #3306 is merged and we can fully test to make sure the elastic terms are assembled properly.

Reply:
hmm, that is interesting. Yes, let's just keep it as it is :)